### PR TITLE
Fix segmented_reduce on empty column with non-empty offsets

### LINKED ISSUE #10556: [BUG] listReduce on a single row input with an empty list generates a zero row output
**Describe the bug**
While testing `ArrayExists` in spark-rapids we hit an exception when a partition consists of a single row with an empty list. The root cause is somewhere behind the fact that listReduce produces an empty output instead of maintaining the invariant: output row count equals input row count 

**Steps/Code to reproduce bug**

The repro via JVM in Scala REPL is as follows.
```scala
val dt = new HostColumnVector.ListType(true, new HostColumnVector.BasicType(true, DType.BOOL8))
val singleEmptyListCV = ColumnVector.fromLists(dt, java.util.Arrays.asList())
// singleEmptyListCV: ai.rapids.cudf.ColumnVector = ColumnVector{rows=1, type=LIST, nullCount=Optional.empty, offHeap=(ID: 5 7fb64ad20650)}
val singleEmptyListReducedCV = singleEmptyListCV.listReduce(SegmentedReductionAggregation.any(), NullPolicy.INCLUDE, DType.BOOL8)
// singleEmptyListReducedCV: ai.rapids.cudf.ColumnVector = ColumnVector{rows=0, type=BOOL8, nullCount=Optional.empty, offHeap=(ID: 6 7fb5b2469d20)}
val gpuCVClz = ShimLoader.loadClass("com.nvidia.spark.rapids.GpuColumnVector")
MethodUtils.invokeStaticMethod(gpuCVClz, "debug", "bad", singleEmptyListReducedCV)
// GPU COLUMN bad - NC: 0 DATA: null VAL: null
// COLUMN bad - BOOL8
```

The output is as expected if the input includes another row with a non-empty list
```scala
val emptyAndNonEmptyListCV = ColumnVector.fromLists(dt, java.util.Arrays.asList(), java.util.Arrays.asList(false))
// emptyAndNonEmptyListCV: ai.rapids.cudf.ColumnVector = ColumnVector{rows=2, type=LIST, nullCount=Optional.empty, offHeap=(ID: 13 7fb5a8171ce0)}
val emptyAndNonEmptyListReducedCV = emptyAndNonEmptyListCV.listReduce(SegmentedReductionAggregation.any(), NullPolicy.INCLUDE, DType.BOOL8)
// emptyAndNonEmptyListReducedCV: ai.rapids.cudf.ColumnVector = ColumnVector{rows=2, type=BOOL8, nullCount=Optional.empty, offHeap=(ID: 14 7fb5b6de1020)}
MethodUtils.invokeStaticMethod(gpuCVClz, "debug", "good", emptyAndNonEmptyListReducedCV)
// GPU COLUMN good - NC: 1 DATA: DeviceMemoryBufferView{address=0x7fb5bfc01000, length=2, id=-1} VAL: DeviceMemoryBufferView{address=0x7fb5bfc01400, length=64, id=-1}
// COLUMN good - BOOL8
// 0 NULL
// 1 false
```

**Expected behavior**
Reduction of a single row with an empty list should produce a single row output 

**Environment overview (please complete the following information)**
 - Environment location: any
 - Method of cuDF install: from source

**Environment details**
TBD

**Additional context**
https://github.com/NVIDIA/spark-rapids/issues/5108

### PR DESCRIPTION
**Pull request by davidwendt**
Fixes `cudf::segmented_reduce` where the input `values` column is empty but the `offsets` are not. In this case, the `offsets` vector `{0,0}` specifies an empty segment which should result in a single null row. The logic has been fixed and new gtest cases have been added.

Closes #10556

### CODE REVIEW

**Review by bdice — APPROVED** [review]
Great! I suppose we tried to be too "clever" with the early exit here and it bit us.

**Review by jlowe — APPROVED** [review]
Tested this with the original Spark query that triggered a failure and it now produces the correct results with this change.  Thanks, @davidwendt!

File: cpp/tests/reductions/segmented_reduction_tests.cpp [hunk 1]
```diff
@@ -387,6 +387,47 @@ TEST_F(SegmentedReductionTestUntyped, ReduceEmptyColumn)
   CUDF_TEST_EXPECT_COLUMNS_EQUAL(*res, expect);
 }
 
+TEST_F(SegmentedReductionTestUntyped, EmptyInputWithOffsets)
+{
+  // values:    {}
+  // offsets:   {0, 0}
```
Comment (jrhemstad) [inline cpp/tests/reductions/segmented_reduction_tests.cpp:393]: Looking through this made me wonder about what happens with:
```
  // values:    {}
  // offsets:   {0, 0, 0, 0, 0, 0}
```
i.e., specifying multiple empty segments. Looking through the code this _should_ still work, but maybe let's modify this test to ensure it does work?

Comment (bdice) [reply cpp/tests/reductions/segmented_reduction_tests.cpp:393]: Is that different than the string test, which includes multiple segments whose offsets are all zero? I thought this case might already be covered. https://github.com/rapidsai/cudf/blob/405a7d3b83fd90287dcf111c9abceb544599fd6f/cpp/tests/reductions/segmented_reduction_tests.cpp#L761

**Review by ttnghia — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**